core/src/main/java/com/netflix/msl/io/UnsynchronizedBufferedInputStream.java

```diff
@@ -0,0 +1,199 @@
+/**
+ * Copyright (c) 2018 Netflix, Inc.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.msl.io;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * <p>A {@code UnsynchronizedBufferedInputStream} adds support for the
+ * {@code mark()} and {@code reset()} functions.</p>
+ *
+ * @author Wesley Miaw <wmiaw@netflix.com>
+ */
+public class UnsynchronizedBufferedInputStream extends FilterInputStream {
```

Comment:
In Java, synchronized blocks that are accessed from a single thread are optimized to not pay the penalty of the semaphore. I think BufferedInputStream the way you were using it was perfectly fine and  would perform identical to this class.

Reply:
If that's the case, why was there a lock in the BufferedReader stack trace? That should have been under single-thread use as well.

Since MSL is security-related, I try to avoid as much as possible pulling in third party libraries. Particularly if the functionality needed is so small and fairly trivial.